### PR TITLE
Add simple "delete to trash" functionality

### PR DESCRIPTION
**Pull request by bauerj**
This pull request has been imported from jonaswinkler/paperless-ng#1249 and was originally opened by tribut on 2021-08-22 21:08:03.

---

Adds `PAPERLESS_THRASH_DIR`. When set, original files are moved there instead of being unlinked when a document is deleted.

Implements the first point from https://github.com/jonaswinkler/paperless-ng/discussions/693.

The idea is that this prevents accidental (or malicious) permanent deletion of a document. This is the most minimalist implementation I could think of, there are no changes to the GUI at all – just a directory that keeps the originals that would otherwise have been removed.

Edit: Failing CI seems bogus (frontend ran into npm timeout and coveralls says coverage decreased by 0.003% - rounding error?)

### CODE REVIEW

**Review by dblitt — CHANGES_REQUESTED** [review]
This pretty much LGTM. I'm not super familiar with the testing, but I'm not seeing any problems. @tribut would you mind putting the `os.rename` call in handlers.py:248 in a try except to handle any exceptions caused by permission errors, etc? And also, there is an extra new line in test_file_handling.py on line 191.

I would still like someone else to review this though I don't expect any problems.

File: src/documents/signals/handlers.py [hunk 1]
```diff
@@ -225,6 +225,28 @@ def set_tags(sender,
...
+                )
+
+                if os.path.exists(new_file_path):
+                    counter += 1
+                else:
+                    break
+
+            logger.debug(
+                f"Moving {instance.source_path} to trash at {new_file_path}")
+            os.rename(instance.source_path, new_file_path)
```
Comment (gador) [inline src/documents/signals/handlers.py:248]: I'd agree and would put it in try, except, too.

**Comment by gador** [comment]
Also, what happens to the files in the folder?
Do we have to manually delete them later on, if we want them gone? 

I'd also suggest adding automatic tests for the newly implemented function.

File: src/documents/signals/handlers.py [hunk 2]
```diff
@@ -225,6 +225,28 @@ def set_tags(sender,
 @receiver(models.signals.post_delete, sender=Document)
 def cleanup_document_deletion(sender, instance, using, **kwargs):
     with FileLock(settings.MEDIA_LOCK):
+        if settings.TRASH_DIR:
+            counter = 0
+            old_filename = os.path.split(instance.source_path)[1]
+            (old_filebase, old_fileext) = os.path.splitext(old_filename)
+
+            while True:
```
Comment (alexander-bauer) [inline src/documents/signals/handlers.py:233]: I think all this logic looks good to me, but I'd definitely like some commentary on this section to make it clearer why we're doing this filename recomposition. (I understand now, but it took me a second!)

**Comment by mweimerskirch** [comment]
I'm not in favour of this. I would prefer implementing a real "history" to be implemented instead. I made a proposal on how I imagine this could be done in the Discussions area here: https://github.com/paperless-ngx/paperless-ngx/discussions/85

**Comment by yschroeder** [comment]
The delete button already exists. Right now it removes files permanently. This PR wants to move them to a trash folder instead.

Also paperless never claimed to be audit proof.

**Comment by Timoms** [comment]
> The delete button already exists. Right now it removes files permanently. This PR wants to move them to a trash folder instead.
> 
> Also paperless never claimed to be audit proof. 

I should really stop commenting when it's 2 am - sorry! But thanks for letting me know!

**Comment by alexander-bauer** [comment]
I've spent some time thinking about this, and wondered if there's room for a broader feature in place of this, to maintain different versions of documents through time, especially in concert with some of the planned editing features.

I think my conclusion is that simpler is better for Paperless, and that new versions of documents go through the ingestion pipeline again, rather than cutting new versions for documents.

So my support is still in on this feature roughly as if.

**Comment by yschroeder** [comment]
Look no further than #85. It is being discussed there

File: src/documents/signals/handlers.py (hunk 2 above)
Comment (tribut) [reply src/documents/signals/handlers.py:235]: I've added a comment on the reason for the file-rename section.

File: src/documents/signals/handlers.py (hunk 1 above)
Comment (tribut) [reply src/documents/signals/handlers.py]: Added.

**Comment by tribut** [comment]
Thanks everyone for the review and the feedback!

I have removed the stray new line and added a comment on the reason for the file-rename section.

I've also wrapped the os.rename in a try block. However, I had initially not done this on purpose: When the user configures this option, they very likely want paperless not to lose a document ever. Which means moving to trash must succeed or we cannot delete the original (i.e. we fail safely). My thinking was that a python exception would be more noticeable by the user than an error message in the log. The best option would of course some sort of prominent error shown in the frontend, but we're not doing this in most other cases, so this is probably fine.

@gador [Tests already exist](https://github.com/tribut/paperless-ng/blob/feature-trash/src/documents/tests/test_file_handling.py#L158), do you have something more in mind?

And yes, currently the files just go to the trash folder without any UI to reach them. I consider this PR an MVP and would welcome some sort of "re-import from trash" function later on. Backing up the current meta-data before deleting would also be nice (see below). But I'd rather get this merged as-is before making this more complex.

@mweimerskirch I disagree that git is a good match for us (I'll comment in #85 later) and think this feature would be orthogonal to a full history, if we decide to include such a feature. In any case removing it later once we have a more complete history functionality that covers this use-case would be fine, I think.

**Comment by gador** [comment]
> @gador [Tests already exist](https://github.com/tribut/paperless-ng/blob/feature-trash/src/documents/tests/test_file_handling.py#L158), do you have something more in mind?

Sorry, I totally missed that. Yes, that's exactly what I suggested 👍

**Comment by dblitt** [comment]
This looks good to me. I agree that deletion shouldn't happen if we can't move the file, and line 257 should handle that fine. Also +1 for backing up the metadata if this is enabled, but I also agree that that is a bigger beast and should be addressed later.

**Review by dblitt — APPROVED** [review]

**Comment by shamoon** [comment]
As per chat, Im going to merge this. I agree its a little weird to have GUI / app put a file where it cant reach it but this essentially seems like an interim fix, and is not enabled by default. When something more comprehensive is done perhaps this should be removed.

**Pull request merged**